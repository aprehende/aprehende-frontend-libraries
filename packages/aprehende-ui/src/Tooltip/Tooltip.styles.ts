import styled from '@emotion/styled';

const tooltipTextColor = '#fff';
const tooltipBackgroundColor = '#000';
const tooltipMargin = '30px';
const tooltipArrowSize = '6px';

export const TooltipWrapper = styled.div`
  display: inline-block;
  position: relative;
`;

export const TooltipContent = styled.div<{
  direction: 'top' | 'bottom' | 'left' | 'right';
}>`
  position: absolute;
  border-radius: 4px;
  left: 50%;
  transform: translateX(-50%);
  padding: 6px;
  color: ${tooltipTextColor};
  background: ${tooltipBackgroundColor};
  font-size: 14px;
  font-family: sans-serif;
  line-height: 1;
  z-index: 100;
  white-space: nowrap;

  &::before {
    content: ' ';
    left: 50%;
    border: solid transparent;
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
    border-width: ${tooltipArrowSize};
    margin-left: calc(${tooltipArrowSize} * -1);
  }

  ${({ direction }) => {
    switch (direction) {
      case 'top':
        return `
          top: calc(${tooltipMargin} * -1);
          &::before {
            top: 100%;
            border-top-color: ${tooltipBackgroundColor};
          }
          `;
      case 'right':
        return `
          left: calc(100% + ${tooltipMargin});
          top: 50%;
          transform: translateX(0) translateY(-50%);
          &::before {
            left: calc(${tooltipArrowSize} * -1);
            top: 50%;
            transform: translateX(0) translateY(-50%);
            border-right-color: ${tooltipBackgroundColor};
          }
        `;
      case 'bottom':
        return `
          bottom: calc(${tooltipMargin} * -1);
          &::before {
            bottom: 100%;
            border-bottom-color: ${tooltipBackgroundColor};
          }
        `;
      case 'left':
        return `
          left: auto;
          right: calc(100% + ${tooltipMargin});
          top: 50%;
          transform: translateX(0) translateY(-50%);
          &::before {
            left: auto;
            right: calc(${tooltipArrowSize} * -2); 
            top: 50%;
            transform: translateX(0) translateY(-50%);
            border-left-color: ${tooltipBackgroundColor};
          }
        `;
    }
  }}
`;
