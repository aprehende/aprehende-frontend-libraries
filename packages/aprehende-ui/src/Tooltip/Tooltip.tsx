import React, { PropsWithChildren, useState } from 'react';

import { TooltipContent, TooltipWrapper } from './Tooltip.styles';

interface TooltipProps {
  delay?: number;
  content: string;
  direction?: 'top' | 'bottom' | 'left' | 'right';
}

const Tooltip: React.FC<PropsWithChildren<TooltipProps>> = ({
  delay,
  content,
  children,
  direction,
}) => {
  let timeout: string | number | NodeJS.Timeout | undefined;
  const [active, setActive] = useState(false);

  const showTip = () => {
    timeout = setTimeout(() => {
      setActive(true);
    }, delay || 400);
  };

  const hideTip = () => {
    clearInterval(timeout);
    setActive(false);
  };

  return (
    <TooltipWrapper onMouseEnter={showTip} onMouseLeave={hideTip}>
      {children}
      {active && (
        <TooltipContent direction={direction || 'top'}>
          {content}
        </TooltipContent>
      )}
    </TooltipWrapper>
  );
};

export default Tooltip;
