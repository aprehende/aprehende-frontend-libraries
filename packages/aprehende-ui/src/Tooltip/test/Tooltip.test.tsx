import { fireEvent, render, waitFor } from '@testing-library/react';
import { describe, it } from 'vitest';

import Tooltip from '../Tooltip';

describe('Tooltip', () => {
  const tooltipContent = 'Tooltip content';
  const elementTrigger = 'Tooltip trigger';

  it('should render correctly', () => {
    const { asFragment } = render(
      <Tooltip content={tooltipContent} delay={0}>
        <div>{elementTrigger}</div>
      </Tooltip>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display tooltip on hover', async () => {
    const { getByText } = render(
      <Tooltip content={tooltipContent} delay={0}>
        <div>{elementTrigger}</div>
      </Tooltip>
    );
    const elementWithTooltip = getByText(elementTrigger);
    fireEvent.mouseEnter(elementWithTooltip);
    await waitFor(() => {
      expect(getByText(tooltipContent)).toBeInTheDocument();
    });
  });

  it('should hide tooltip on mouse leave', async () => {
    const { getByText, queryByText } = render(
      <Tooltip content={tooltipContent} delay={0}>
        <div>{elementTrigger}</div>
      </Tooltip>
    );
    const elementWithTooltip = getByText(elementTrigger);
    fireEvent.mouseEnter(elementWithTooltip);
    await waitFor(() => {
      expect(getByText(tooltipContent)).toBeInTheDocument();
    });
    fireEvent.mouseLeave(elementWithTooltip);
    await waitFor(() => {
      expect(queryByText(tooltipContent)).not.toBeInTheDocument();
    });
  });

  it('should display tooltip in the right position', async () => {
    const { getByText } = render(
      <Tooltip content={tooltipContent} delay={0} direction="right">
        <div>{elementTrigger}</div>
      </Tooltip>
    );
    const elementWithTooltip = getByText(elementTrigger);
    fireEvent.mouseEnter(elementWithTooltip);
    await waitFor(() => {
      expect(getByText(tooltipContent)).toBeInTheDocument();
    });
  });

  it('should display tooltip in the left position', async () => {
    const { getByText } = render(
      <Tooltip content={tooltipContent} delay={0} direction="left">
        <div>{elementTrigger}</div>
      </Tooltip>
    );
    const elementWithTooltip = getByText(elementTrigger);
    fireEvent.mouseEnter(elementWithTooltip);
    await waitFor(() => {
      expect(getByText(tooltipContent)).toBeInTheDocument();
    });
  });

  it('should display tooltip in the bottom position', async () => {
    const { getByText } = render(
      <Tooltip content={tooltipContent} delay={0} direction="bottom">
        <div>{elementTrigger}</div>
      </Tooltip>
    );
    const elementWithTooltip = getByText(elementTrigger);
    fireEvent.mouseEnter(elementWithTooltip);
    await waitFor(() => {
      expect(getByText(tooltipContent)).toBeInTheDocument();
    });
  });
});
