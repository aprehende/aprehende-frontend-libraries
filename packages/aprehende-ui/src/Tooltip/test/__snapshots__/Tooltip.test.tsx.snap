// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1tx6twl"
  >
    <div>
      Tooltip trigger
    </div>
  </div>
</DocumentFragment>
`;
